### PR TITLE
Using Kotlin's structured concurrency model

### PR DESCRIPTION
**Pull request by florina-muntenescu**
ViewModels have a scope and the parent job gets cancelled in `onCleared`
Repositories and UseCases, that are still called from Java and use the mix of coroutines and callbacks, since they also offer a way of cancelling the tasks, are using the `GlobalScope` to launch coroutines.

### CODE REVIEW

File: core/src/main/java/io/plaidapp/core/designernews/domain/LoadStoriesUseCase.kt [hunk 1]
```diff
@@ -44,16 +45,16 @@ class LoadStoriesUseCase(
         page: Int,
         callback: LoadSourceCallback,
         jobId: String
-    ) = launch(contextProvider.io) {
+    ) = GlobalScope.launch(dispatcherProvider.computation) {
```
Comment (chrisbanes) [inline core/src/main/java/io/plaidapp/core/designernews/domain/LoadStoriesUseCase.kt]: You probably don't want GlobalScope. Wrap it in a `coroutineScope {}` call.

File: core/src/main/java/io/plaidapp/core/designernews/domain/SearchStoriesUseCase.kt [hunk 2]
```diff
@@ -44,16 +45,16 @@ class SearchStoriesUseCase(
         page: Int,
         callback: LoadSourceCallback,
         jobId: String
-    ) = launch(contextProvider.io) {
+    ) = GlobalScope.launch(dispatcherProvider.computation) {
```
Comment (chrisbanes) [inline core/src/main/java/io/plaidapp/core/designernews/domain/SearchStoriesUseCase.kt]: Ditto about GlobalScope

File: designernews/src/main/java/io/plaidapp/designernews/ui/login/LoginViewModel.kt [hunk 3]
```diff
@@ -54,36 +58,39 @@ class LoginViewModel(
...
-        val result = loginRepository.login(username, password)
-
-        if (result is Result.Success) {
-            val user = result.data
-            emitUiState(
-                showSuccess = Event(
-                    LoginResultUiModel(
-                        user.displayName.toLowerCase(),
-                        user.portraitUrl
+    private fun launchLogin(username: String, password: String) =
```
Comment (chrisbanes) [inline designernews/src/main/java/io/plaidapp/designernews/ui/login/LoginViewModel.kt]: Nit: use block style if the line goes over 1 line

Comment (florina-muntenescu) [reply designernews/src/main/java/io/plaidapp/designernews/ui/login/LoginViewModel.kt]: done

File: core/src/main/java/io/plaidapp/core/designernews/domain/SearchStoriesUseCase.kt (hunk 2 above)
Comment (florina-muntenescu) [reply core/src/main/java/io/plaidapp/core/designernews/domain/SearchStoriesUseCase.kt]: done

**Comment by florina-muntenescu** [comment]
@chrisbanes reading for another 👀

File: build.gradle [hunk 4]
```diff
@@ -23,7 +23,7 @@ buildscript {
             'androidx'          : '1.0.0',
             'constraintLayout'  : '2.0.0-alpha2',
             'coreKtx'           : '1.0.0',
-            'coroutines'        : '0.26.1-eap13',
+            'coroutines'        : '0.30.0-eap13',
```
Comment (chrisbanes) [inline build.gradle]: There's 0.30.2-eap13 now (or even 1.0.0-RC1)

File: core/src/main/java/io/plaidapp/core/dribbble/data/ShotsRepository.kt [hunk 5]
```diff
@@ -63,13 +64,13 @@ class ShotsRepository constructor(
         page: Int,
         id: String,
         onResult: (Result<List<Shot>>) -> Unit
-    ) = launch(contextProvider.io) {
+    ) = GlobalScope.launch(dispatcherProvider.io) {
```
Comment (chrisbanes) [inline core/src/main/java/io/plaidapp/core/dribbble/data/ShotsRepository.kt]: Sure you want to use GlobalScope? The standard way would be to do:

``` kotlin
private fun launchSearch(...) = coroutineScope {
    launch(dispatcherProvider.io) {
        // ...
    }
}
```

Comment (florina-muntenescu) [reply core/src/main/java/io/plaidapp/core/dribbble/data/ShotsRepository.kt]: `coroutineScope` doesn't work with the mix of coroutines and callbacks we have here (due to `search` being called from Java). 
But, I replaced the GlobalScope with one defined in this class.

**Pull request merged**